databricks_connector.py

```diff
@@ -0,0 +1,715 @@
+                container["artifacts"] = [{"cef": alert.as_dict()}]
+                self.save_container(container)
+
+                self._state["alerts"] = state_alerts
+
+        return action_result.set_status(phantom.APP_SUCCESS)
+
+    def handle_action(self, param):
+        # Get the action that we are supposed to execute for this App Run
+        action_id = self.get_action_identifier()
+
+        self.debug_print("action_id", self.get_action_identifier())
+
+        try:
+            if action_id == "test_connectivity":
+                ret_val = self._handle_test_connectivity(param)
+            elif action_id == "list_alerts":
+                ret_val = self._handle_list_alerts(param)
+            elif action_id == "list_clusters":
+                ret_val = self._handle_list_clusters(param)
+            elif action_id == "create_alert":
+                ret_val = self._handle_create_alert(param)
+            elif action_id == "delete_alert":
+                ret_val = self._handle_delete_alert(param)
+            elif action_id == "get_job_run":
+                ret_val = self._handle_get_job_run(param)
+            elif action_id == "get_job_output":
+                ret_val = self._handle_get_job_output(param)
+            elif action_id == "perform_query":
+                ret_val = self._handle_perform_query(param)
+            elif action_id == "get_query_status":
+                ret_val = self._handle_get_query_status(param)
+            elif action_id == "cancel_query":
+                ret_val = self._handle_cancel_query(param)
+            elif action_id == "execute_notebook":
+                ret_val = self._handle_execute_notebook(param)
+            elif action_id == "list_warehouses":
+                ret_val = self._handle_list_warehouses(param)
+            elif action_id == "on_poll":
+                ret_val = self._handle_on_poll(param)
+            else:
+                action_result = self.add_action_result(ActionResult(dict(param)))
+                ret_val = action_result.set_status(
+                    phantom.APP_ERROR,
+                    consts.UNHANDLED_ACTION_ID_ERROR_MESSAGE.format(action_id),
+                )
+        except Exception as e:
+            error_message = self._get_error_msg_from_exception(e)
+            action_result = self.add_action_result(ActionResult(dict(param)))
+            ret_val = action_result.set_status(phantom.APP_ERROR, error_message)
+
+        return ret_val
+
+    def initialize(self):
+        # Load the state in initialize, use it to store data
+        # that needs to be accessed across actions
+        self._state = self.load_state()
+
+        # get the asset config
+        config = self.get_config()
+
+        self._host = config["host"]
+        self._username = config.get("username")
+        self._password = config.get("password")
+        self._token = config.get("token")
+
+        if not (self._username and self._password) and not self._token:
```

Comment:
Why?

While the two are logically equivalent, I think the `not (... and ...)` is more intuitive. "If the user didn't provide a (username AND password), AND they didn't provide a token...."

Reply:
I'm very sure that in the long run applying demorgan's laws is more readable, which is very clear in more complicated cases, but even in simpler ones like this one it's just easier for a person to parse what has to be true instead of adding that extra layer of `not`. And more often than not the intuitiveness comes from the fact that you wrote this code and worked on it for a while and so you know from just a glance what to expect.

Reply:
Refactored it to check the positive case instead of the negative, which is even more readable.